### ISSUE TITLE
Support for other pam modules on Swaylock

### ISSUE THREAD

**Issue by hkupty** [swaywm/sway#1832]
Hi! I know this is low priority compared to the amazing work you're doing on moving sway to wlroots, but one of the things that'd hold me back from moving to sway is the lack of support for MFA on Swaylock.
I use a yubikey as a second factor of authencication and TOTP (google auth) as a fallback in case I lose the yubikey. For some reason, it seems that neither scenarios are covered or supported by swaylock.

Anyway, thanks for the good work.

Cheers,
Henry John Kupty

**Comment by martinetd** [comment]
You can write whatever you want in /etc/pam.d/swaylock, it is standard pam so should do anything you ask it to (defaults to the 'login' ones so your setup might work out of the box if you have configured it globally; if you haven't it's just a file to change)

**Comment by hkupty** [comment]
Here is my `/etc/pam.d/swaylock`:
```
auth   required    pam_unix.so
auth   sufficient    pam_yubico.so mode=challenge-response ...
auth   required    pam_google_authenticator.so
```

This exactly the same I have for my `physlock` and my `sudo` setups. Both of them act as I described:
> I use a yubikey as a second factor of authencication and TOTP (google auth) as a fallback 

What happens is that swaylock doesn't support any of those, as I described:
> For some reason, it seems that neither scenarios are covered or supported by swaylock

In details, it fails with that pam setting, either if I have the yubikey plugged in (which would suffice since it is `sufficient` on this stage) or if it is supposed to prompt me for the TOTP code.

The thing is, it seams none of the following modules are called and it fails with the password. If I was mistyping my password, it'd fail after prompting me for the TOTP code, which is not happening.

**Comment by martinetd** [comment]
Ah, right, totp requires displaying something so that'll need extra support, I didn't pay attention to the last bit.
Challenge-response shouldn't need anything though; the yubico website says there's a debug mode so I suppose that might help you figure out what goes wrong...

**Comment by SirCmpwn** [comment]
If you think this should be a priority, implement it and send a patch.

**Comment by hkupty** [comment]
Ok, so I figured out what happens with the yubikey. The challenge-response file is stored under `/var/yubico` accessible only by root. As described on their [documentation](https://developers.yubico.com/yubico-pam/Authentication_Using_Challenge-Response.html):
```
sudo mkdir /var/yubico
sudo chown root.root /var/yubico
sudo chmod 700 /var/yubico
```

This means that only a process being run as root can check if the attached yubikey is the one that generated that response for that challenge.

Swaylock is installed as root, but is run as the user which calls it. It should have `setUid` bit set for it, in order to inherit root access and access yubikey. This is what `physlock` does.

Now, for TOTP, I'm not sure how that prompt would need to be implemented.

**Comment by SirCmpwn** [comment]
Running swaylock as root is an incredibly stupid idea.

**Comment by hkupty** [comment]
How else would you get `/var/yubico` data?

**Comment by SirCmpwn** [comment]
I'm not adding yubico-specific code to swaylock.

**Comment by hkupty** [comment]
I'm not asking that. That is done by its pam module. The thing is, having pam to manage privilege access is a way to ensure setuid is secure. They're related on that end, otherwise `swaylock` is just a fancy screensaver that has no actual security over the machine, such as blocking tty.

The thing, in the end, is that if swaylock isn't run by root, it can't block that, so I can easily change the tty and `pkill swaylock`, which was exactly what I did when I got stuck by swaylock not being able to run the pam modules it should.

**Comment by SirCmpwn** [comment]
>so I can easily change the tty and `pkill swaylock`

Don't leave yourself logged in on other ttys, then? You still have to enter a password at getty unless you configure it to accept no password, in which case I can just reboot your computer and get in anyway.

Even so, swaylock doesn't have to be root to prevent tty switching. Sway can already do that.

**Comment by emersion** [comment]
>so I can easily change the tty and pkill swaylock

You could also just kill sway directly.

**Comment by hkupty** [comment]
> Sway can already do that.

If that is the case, then it's buggy, because I've being swapping back and forth between a sway session to test this.

Even so, you're relying on 'swaylock users will have their ttys secure' to make sure swaylock is secure.

But we're diverging from the topic.

It'd be totally fine if you otherwise state that you don't support pam, but most pam modules require to actually be run as root. If you're breaking that invariant, then you're making most of pam security useless, thus not entirely supporting it.

**Comment by SirCmpwn** [comment]
Sway *can* do this, not that it does do it.

PAM works fine now without root. I'm not making it root.

**Comment by hkupty** [comment]
I'll give sway another shot. I do believe in its potential, so the issue still remains - mfa support is missing from swaylock.

Even though there may be other ways to deal with dealing with challenge-response as in the yubikey - and I'll not discuss this for now - , there are other forms of mfa which might require another password to be typed and that is not supported by swaylock, which fails right after the password was correctly typed.

**Comment by SirCmpwn** [comment]
Patches welcome, but I haven't changed my mind about swaylock running as a normal user.

**Comment by hkupty** [comment]
I'm not questioning the user running swaylock - though I'd appreciate a debate over the theme. Still, I think there *should* be a way to deal with MFA, which doesn't need to be a challenge-response to a third-party hardware.

I'll see what - and if - I can do.

**Comment by divVerent** [comment]
> Running swaylock as root is an incredibly stupid idea.

How else are you gonna authenticate against the system password on FreeBSD? There you _must_ be root to do that, even with PAM.

To solve this problem in xsecurelock, I had to factor out the talking to PAM into a separate process (called `authproto_pam` there), so I can run PAM as setuid root, while having anything UI run as the calling user. Was annoying to implement, but still much better than what everyone else does (namely, run the entire screen lock as root).

**Comment by ddevault** [comment]
swaylock can fork, drop root on the parent, and use the child to call `getspnam`, which works on FreeBSD. It then drops root and validates passwords over a pipe.

I don't want to complicate the root code here by introducing PAM to the equation. Right now it's small, isolated, and only has root for a short period of time. It's easily reasoned about and easily verified.

**Comment by divVerent** [comment]
Yeah, that is the non PAM way though, and PAM is pretty important e.g. for
2FA stuff.

Of course, if you asked me this is a bug in PAM - shouldn't there be a way
I make PAM open all file handles it needs, then drop privs? But there is
none.
